Translate uninstaller to Bash

gromgit opened the pull request:
For clearer 1-to-1 correspondence with the original Ruby code,
I've adopted the naming convention:

  Class::method -> Class_method()

dawidd6 commented on uninstall.sh:
```diff
@@ -0,0 +1,348 @@
+}
+
+String_strip() {
+  local s; for s in "$@"; do
```
`local s` might be better to be in a separate line. Readability and consistency with `Dir_children`.

gromgit replied:
Good point, will fix.

dawidd6 commented on uninstall.sh:
```diff
@@ -0,0 +1,348 @@
+
+Array_shell_s() {
+  local result=("$1"); shift
+  local s; for s in "$@"; do
```
Same as above.

gromgit replied:
Good point, will fix.

gromgit commented:
Clarification request from a complete Ruby newbie. In the following code from the Ruby `uninstall`:
```ruby
homebrew_files.each do |file|
  if options[:dry_run]
    puts "Would delete #{file}"
  else
    begin
      FileUtils.rm_rf(file)
    rescue => e
      warn "Failed to delete #{file}"
      puts e.message
      @failed = true
    end
  end
end
```
Can `FileUtils.rm_rf()` throw an exception? If I'm reading [the relevant RubyDoc](https://ruby-doc.org/stdlib-2.4.1/libdoc/fileutils/rdoc/FileUtils.html#method-c-rm_rf) correctly, it never does, which explains a significant difference with the bash equivalent:
```sh
for file in "${homebrew_files[@]}"; do
  if [[ $opt_dry_run ]]; then
    echo "Would delete ${file}"
  else
    err=$(rm -fr "$file" 2>&1)
    if [[ $err ]]; then
      warn "Failed to delete ${file}"
      echo "$err"
      failed=true
    fi
  fi
done
```
When both instance try to remove a protected directory like `/usr/local/Cellar`, the Ruby code seems to silently ignore the failure, while the bash code registers an `rm` "Permission denied" failure.

itchyny commented on uninstall.sh:
```diff
@@ -0,0 +1,350 @@
+  rm -fr "$tmpdir"
+  # Invalidate sudo timestamp before exiting
+  /usr/bin/sudo -k
+' exit
```
I think it's common to use UPPER CASE for traps.

gromgit replied:
I've seen it both ways in about equal measure, but OK, I'll fix that.

itchyny commented on uninstall.sh:
```diff
@@ -0,0 +1,350 @@
+Tty_blue=$(Tty_mkbold 34)
+Tty_red=$(Tty_mkbold 31)
+Tty_bold=$(Tty_mkbold 39)
+Tty_reset=$(Tty_escape 0)
```
I believe case consistency with installer.sh matters.

gromgit replied:
OK. As I mentioned in the commit text, this was to parallel "the Ruby way" for easier translation, but I'll change that.

MikeMcQuaid replied:
Thanks both of you. Agreed that consistency with install.sh would be good for now.

itchyny commented on uninstall.sh:
```diff
@@ -0,0 +1,350 @@
+    result+=("${s// /\\ }")
+  done
+  echo -n "${result[*]}"
+}
```
Why is this implementation different from `shell_join` in installer.sh?

gromgit replied:
Mostly because it didn't click in my mind that they were actually the same function. Will replace with `shell_join`.

gromgit replied:
Along with any other "twinned" functions.

MikeMcQuaid replied:
Agreed that consistency with install.sh would be good for now.

itchyny commented:
Do you have plan to update the `uninstall` script to use `uninstall.sh` like the installer? After that, I think we can remove the Ruby things (Gemfile, Rubocop, CI), but I'd like to opinions from main maintainers.

itchyny commented on uninstall.sh:
```diff
@@ -0,0 +1,350 @@
+fi
+
+Dir_children "$HOMEBREW_REPOSITORY" "$HOMEBREW_PREFIX" \
+| sort -u > "$tmpdir/residual_files"
```
Let's add indentation here.

gromgit replied:
Sorry, where exactly?

MikeMcQuaid replied:
```suggestion
  | sort -u > "$tmpdir/residual_files"
```
presumably

gromgit replied:
OK, done.

itchyny commented on uninstall.sh:
```diff
@@ -0,0 +1,350 @@
+
+if [[ ${#homebrew_prefix_candidates[@]} -eq 0 ]]; then # Attempt to locate Homebrew unless `--path` is passed
+  prefix=$(brew --prefix)
+  [[ $prefix ]] && homebrew_prefix_candidates+=("$prefix")
```
I think it's preferable to write `-n` explicitly.

gromgit replied:
OK.

gromgit commented:
> Do you have plan to update the `uninstall` script to use `uninstall.sh` like the installer? After that, I think we can remove the Ruby things (Gemfile, Rubocop, CI), but I'd like to opinions from main maintainers.

Good point. I'd not actually looked at `install`, but yeah, `uninstall` should probably do the same.

MikeMcQuaid commented:
> Do you have plan to update the `uninstall` script to use `uninstall.sh` like the installer? After that, I think we can remove the Ruby things (Gemfile, Rubocop, CI), but I'd like to opinions from main maintainers.

Yep, also think this would be good.

gromgit commented:
OK, made all the suggested changes, including replacing `uninstall` with its shim variant.

itchyny commented:
Would you change the p/r title contains the word uninstaller?

gromgit commented:
> Would you change the p/r title contains the word uninstaller?

Done.

itchyny commented on uninstall.sh:
```diff
@@ -0,0 +1,388 @@
+    homebrew_prefix_default=/usr/local
+    # Stock macOS doesn't have a realpath, so we make do with builtin Python
+    realpath() {
+      /usr/bin/python -c 'import os,sys;print(os.path.realpath(sys.argv[1]))' "$1"
```
I'm afraid introducing dependency on _builtin Python_ is not a good option. Translation to Bash started from the deprecation of scripting languages (see [49764202](https://developer.apple.com/documentation/macos_release_notes/macos_catalina_10_15_release_notes)). I think Python will be installed while installing the command line tools but is it guaranteed to keep it after uninstallation? @MikeMcQuaid What do you think?

gromgit replied:
> [49764202](https://developer.apple.com/documentation/macos_release_notes/macos_catalina_10_15_release_notes))

Oh. That's not good. Will find a pure-bash replacement.

MikeMcQuaid replied:
agreed with @itchyny. This is the method I tend to use: https://stackoverflow.com/a/35996728/5355

gromgit replied:
OK, here's the version I'm inclined to go with. It takes care not to output anything except the canonicalized pathname, or nothing at all if it can't be canonicalized for whatever reason, especially link chains 100 or more levels deep (usually a symlink cycle):
```sh
realpath() {
  local dir base newpath
  dir=$(dirname "$1")
  base=$(basename "$1")
  (
    cd -P "$dir" &> /dev/null || exit 1
    redirs=0
    while [[ -L $base && $((redirs++)) -lt 100 ]]; do
      newpath=$(readlink "$base" 2>/dev/null) || exit 1
      dir=$(dirname "$newpath")
      base=$(basename "$newpath")
      cd -P "$dir" &> /dev/null || exit 1
    done
    if [[ -L $base ]]; then
      exit 1
    else
      echo "$(pwd -P)/$base"
    fi
  )
}
```
Comments?

MikeMcQuaid replied:
That seems excessively complicated. `cd` and `pwd` with an argument should be sufficient.

gromgit replied:
That would only work if the entire path points to an existing directory, so it'll fail when we test, say, `$HOMEBREW_PREFIX/bin/brew`, as is already done in the uninstaller.

MikeMcQuaid replied:
If it doesn't exist it can/will return an empty string which can itself be filtered out.

gromgit replied:
Are you talking about:
```
cd "$(dirname "$target")" && echo "$(pwd -P)/$(basename "$target")"
```
? I guess that might work.

MikeMcQuaid replied:
Aye, thanks!

itchyny commented on uninstall.sh:
```diff
@@ -0,0 +1,388 @@
+  prefix=$(brew --prefix)
+  [[ -n $prefix ]] && homebrew_prefix_candidates+=("$prefix")
+  # shellcheck disable=SC2230  # "which" is a legitimate fallback
+  prefix=$(command -v brew) || prefix=$(which brew) || prefix=""
```
I see this is a direct translation from the original code, but I'm not sure we still need this fallback code (`command -v brew` fails but `which brew` succeed?). In ruby it invokes the executable (maybe `/usr/bin/command`), but in bash it's a builtin function.

gromgit replied:
Good point, will fix.

gromgit commented:
Very sorry for the long delay, had to deal with some personal issues. I've fixed everything previously discussed, and also squashed a bug related to cleaning up Info files.

MikeMcQuaid commented:
No worries @gromgit, hope you and your family are all OK. Any thoughts @itchyny @dawidd6?

MikeMcQuaid commented:
Merged, thanks @gromgit! Would welcome more PRs from you or anyone else to clean up and make the installer/uninstaller (somehow?) share code and generally be more consistent.
